Marketplace simulation tutorial + ficcional inputs

DiSanchz opened the pull request:
#WIP: This pull request contains that the most recent advances on the Marketplace simulation tutorial notebook. It was build with a very similar structure compared the the non-mp tutorial notebook and also under the assumption that the user have visited it before and is somewhat familiar with non-mp simulations. All in all, though it provides some context, it is mostly focused on the practical aspects of running the simulation as in the case of the non-mp tutorial notebook. E.g. I did not dig too deep on the details of the ML tasks but just presented them as it was discussed in the last meeting. 

Besides the notebook, the ficcional inputs for the ML tasks is also included in the pull request (productspace.tsv, userspace.tsv & rating_histogram_all). I was not sure if you expected these to be part of the pull request but finally decided to included so you can check their content. They include information on the embeddings and ratings of 1400 products & users, this number is arbitrary, I choose it as I did not consider it too large for running in my own machine and also because I wasn't sure either if you wanted me to replicate the exact same dimensions of the original inputs contained in the bucket storage. In any case, if required I can do this pretty easily given that I already have the code to produce these replica input files with any number of dimensions. Any other change on how these are generated can also be easily implimented if needed.

All the code in the notebook works, including the simulation and the training/fitting of the models when given the ficcional inputs.

Looking forward on feedback about any addition, modification, deletion etc.!

@narendramukherjee 
@shrabasteebanerjee 

PD: Git hub warns that the pull request has conflicting points with the main repo on the non-marketplace-simulations tutorial notebook. I belive this must be related to one or two minor typos I discovered and corrected in that notebook while preparing the other content included in the pull request.

DiSanchz committed:
non_mp_notebook_draft_added

DiSanchz committed:
Non-MP_tutorial_updated

DiSanchz committed:
Non-MP_tutorial_updated_II

DiSanchz committed:
Non-MP_tutorial_updated_III

DiSanchz committed:
marketplace_tut_plus_inputs

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,682 @@
+   "id": "9b21f40f",
+   "metadata": {},
+   "source": [
+    "Another option to adjust is the number of CPUs involved in the simulation. The cell below is set so all the CPUs available are employed in the simulation. (Note: The MarketplaceSimulator class, which is in charge of the simulation, is designed to paralellize the simulation of the different marketplaces across all the CPUs to be employed.)"
```
The CPU number adjustment doesn't happen through Pytorch for the simulations - this method of adjusting CPU numbers to be used is actually for the inference part where Pytorch is used. We can probably remove this from the simulation notebooks and just keep it in the inference notebook

DiSanchz replied:
Done!

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,682 @@
+   "id": "24787404",
+   "metadata": {},
+   "source": [
+    "Beyond the overall scheme presented in the introduction, markeplace simulations incorporate two other significant new elements which were not present in the non-marketplace setting and that deserve to be mentioned before actually runing the simulation. The first of these are the Gaussian mixture models (GMM) fitted on the products' and users' embeddings that are then used to draw the embeddi ...
```
Can you just put these 2 things as numbered/bullet points 1 and 2?

DiSanchz replied:
Done!

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,682 @@
+    "\n",
+    "At this point, if you intend to run an example marketplace simulation you can just carry on with the rest of the notebook given that example data for such task is already provided. However if you are interested in running the simulation with your own data on products and users you should consider that the processes mentioned above require the following inputs:\n",
+    "\n",
+    "1. A file called \"productspace.tsv\" containing the 100-dimensional product embeddings, each one accompained by a product identifier in the format \"product_< id >\" where the id is a 7 digit number (e.g. product_1234567)\n",
```
Product ID doesn't have to be 7 digit actually - it is only 7 digit in our dataset but might not be in other datasets. Just product_<unique id of product> will be enough. We should just say that the id should not contain underscores (_) as then the later code will try to break up the string by underscores and will not recover the right ids if extra underscores are there

DiSanchz replied:
Solved!

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,682 @@
+    "3. A third file named \"rating_histogram_all.txt\" containing the numerical IDs and actual rating distribution [1 - 5] of the products in \"productspace.tsv\".\n",
+    "\n",
+    "\n",
+    "These files need to be placed in the \"snpe/artifacts/marketplace/\" path within the repository. For further clarity on the exact format expected for the inputs you can visit the example versions of these already provided at the aforementioned path. Once the inputs have been provided as specified the code cells below headings B.1 and B.2 can be executed to fit the GMM models for embeddings a ...
```
This artifact path requirement is not strict - if you use this default path, GMM and the NN work without any artifact path being supplied as parameter. But you can also give artifact_path as an argument to both these classes: https://github.com/narendramukherjee/reputation-systems/blob/master/snpe/snpe/embeddings/embeddings_density_est_GMM.py#L16

DiSanchz replied:
Perfect, I've inlcuded this information in the notebook!

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,682 @@
+    "- __num_products__: Number of products to be included in each marketplace. For example, if set at a value of 1400, a total of 1400 different virtual products will be sampled from the GMM fit of the provided product embeddings for each of the simulated marketplaces. \n",
+    "\n",
+    "\n",
+    "- __num_marketplace_reviews__: Desired number (Integer) of ratings to be obtained to conclude the simulation of each marketplace. Alternatively the simulation of a marketplace will also be concluded automatically if this number of ratings has not been reached by the time that an amount of consumers 30 times larger have visited the marketplace. E.g. In a case where this argument is set to 10, ...
```
Just clarify that this is the total number of reviews added across all the products in the marketplace. This is important because for non-marketplace simulations, we specify the number of reviews we want per product and end the simulation when that product-specific number is hit, but of course, that we can't do for the marketplace simulations

DiSanchz replied:
Done!

narendramukherjee reviewed (changes requested):
@DiSanchz This is really good work! I have added a few minor comments to be addressed. And only 1 big comment: can you add the code you used to generate the fictional input to the GMM and the EmbeddingRatingPredictor models? 

I agree with adding the fictional input files themselves as you said - will be very helpful to the end user to understand the kind of input we want from them.

DiSanchz committed:
input_generation_added

DiSanchz committed:
input_generation_added_2

DiSanchz commented:
@narendramukherjee I've just added to the notebook the code regarding input generation. For it to be fully functional I've also needed to include in this last commit the "anonimized" rating histogram file as discussed in the meeting from two weeks ago. This file is not technically a tutorial notebook so maybe it could be moved to somewhere  else (e.g. a sub-directory) but for proximity I though it would also make sense to keep it close to the markeplace simulation notebook tutorial. Besides that I've also added a few minor improvements to the notebooks (e.g. the "output" directory meant to contain tutorials' simulational output in a more organized way).

On top of the general feedback regarding the notebook and the last commit, In this case I would also greatly appreciate any general suggestion regarding the coding practices I employ in this (or any) snippet and ways in which you feel It could be improved or make it more professional-looking.

DiSanchz committed:
generation_code_added_3

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,957 @@
+    "\n",
+    "At this point, if you intend to run an example marketplace simulation you can just carry on with the rest of the notebook given that example data for such task is already provided. However if you are interested in running the simulation with your own data on products and users you should consider that the processes mentioned above require the following inputs:\n",
+    "\n",
+    "1. A file called \"productspace.tsv\" containing the 100-dimensional product embeddings, each one accompained by a product identifier in the format \"product_< id-number >\" (e.g. product_1234567). Additionally, make sure that the id-number itself does not contain any underscore itself as this may cause problem in the input processing stage.\n",
```
Actually, the embeddings don't need to be 100-dimensional strictly. They can be any dimension size, just that (for obvious reasons) the product and user embeddings should have the same dimensionality

DiSanchz replied:
Perfect! I've sustituted the mentions "100-dimensional" to "n-dimensional"

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,957 @@
+    "\n",
+    "- __num_simulations__: Total number (Integer) of marketplaces to be simulated. Each marketplace will be composed of its unique set of virtual products and consumers sampled from the GMM fit of the provided product embeddings.\n",
+    "\n",
+    "The generate_and_save_simulations function defined below is in charge of calling the MarkeplaceSimulator class and deliver the required arguments to configure and carry out the simulation. In case you would like to re-visit the interpretations of the other arguments which are non-idiosyncratic of markeplace simulations you can scroll further  down the notebook where a brief review has been i ...
```
What do you mean by "non-idiosyncratic" here? Do you mean to say that these arguments come from the non-marketplace simulator and are not newly introduced by the marketplace setting? If yes, don't use idiosyncratic as it is interpreted a little differently, but just say what I said in the previous sentence

DiSanchz replied:
That is exaclty what I meant, I've modified te fragment to substitute the "non-idiosyncratic" expression

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,957 @@
+    }
+   ],
+   "source": [
+    "generate_and_save_simulations(16, np.ones(5), 0.05, \"timeseries\", \"mode\", 5, 10, 5, 3)"
```
Actually I see an error block in the notebook through a keyboard interrupt. Can we just let the simulation complete its run to produce reviews for a small number of products? Just do a small number of marketplaces like 4 or 8 with a handful of products (like 10 like you chose) and do say 200 reviews per marketplace to ensure that all products get some reviews

DiSanchz replied:
Done!

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,957 @@
+    "    to look for the inputs by default.\n",
+    "    A different path storage can be provided as the value of the optional \"artifact_path\" argument. \n",
+    "    \"\"\"\n",
+    "    assert (\n",
```
Why this very specific number of products?

DiSanchz replied:
This is because 13939 is the number of histograms included in the rating_histogram.txt file, any larger number should produce an error as there are not enough rows in the dataframe (i.e. review histograms) to assign to the generated product IDs if the number of these is larger than that.

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,957 @@
+    "    ids_product = []\n",
+    "\n",
+    "    # Generating a set of random product ids to be used in rating histograms and embeddings\n",
+    "    while len(ids_alone) < number_prods:\n",
```
So here you can just do a choice without replacement in a single line of code to get unique ids like:
```
np.random.choice(np.arange(1000000, 9999999), size=number_prods, replace=False)
```
Usually using numpy arrays and their associated methods and manipulations also produces code that is MUCH faster than list comprehension written in pure Python

DiSanchz replied:
Got it! I'll try to use np arrays from now on when possible! I knew the very basics of numpy before though hadn't use it a lot beyond that. Luckily this exact week a few days after I wrote this code I've had a lecture about numpy as part of my master's and that has definitively helped!

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,957 @@
+    "            ids_alone.append(a)\n",
+    "            ids_product.append(\"product_\" + str(a))\n",
+    "\n",
+    "    # Generating 100d arrays to be used as product embeddings\n",
```
Once again use numpy :) 
To generate 100-D arrays of random numbers, you can just do something like:
`np.random.normal(size=(number_prods, 100))`
Drawing random numbers from the normal distribution here is more appropriate than uniform (which is what random.random does) as these embeddings have values in (-inf, +inf) and not in [0,1] like uniform.

DiSanchz replied:
Got it!

narendramukherjee commented on snpe/tutorials/marketplace_simulation.ipynb:
```diff
@@ -0,0 +1,957 @@
+    "    for i in range(number_users):\n",
+    "        hundred_space_users.append(list((random.rand(100) * 2) - 1))\n",
+    "\n",
+    "    # Including product ids alogside product embeddings\n",
```
If you have these embeddings as numpy arrays, you can use the pandas function to convert np arrays into DF: `df = pd.DataFrame(my_array, columns = ['Column_A','Column_B','Column_C'])`
https://datatofish.com/numpy-array-to-pandas-dataframe/

DiSanchz replied:
Done!

narendramukherjee reviewed (changes requested):
Great work @DiSanchz ! Only very minor additional comments on the old stuff plus some comments specific to the code that generates the artifical input for GMM and NN. 

Unrelated to all this, if you find a numpy tutorial at some point, go through it. Pytorch is also very similar in syntax to numpy, so it'll be useful that way too. Generally, numpy arrays need a bit of different thinking than pandas DFs, and are a great tool to be familiar with. Don't worry, you'll see enough use of them in my coding through this project :)

DiSanchz committed:
marketplace_last_mods

DiSanchz committed:
some_minor_fixes

DiSanchz committed:
unnecesary_tsv_txt_deleted

DiSanchz committed:
Merge branch 'master' into Marketplace_simulation_tutorial

DiSanchz commented:
Conflicts resolved! PR ready to merge! @narendramukherjee 

PD: After the last modifications GitHub is telling me that there are 'changes requested', however, this is stated in a bolder way than what I've previously seen in past PRs. I understand that this showier way to warn me of these requests responds to the fact that I've declared the PR as ready to merge from its previous draft status (which is something I have not done in previous PRs). All requested changes have been already addressed.

DiSanchz committed:
unrelated_files_deleted

DiSanchz commented:
Pull request ready to merge! 

PD: After the last modifications GitHub is telling me that there are 'changes requested', however, this is stated in a bolder way than what I've previously seen in past PRs. I understand that this showier way to warn me of these requests responds to the fact that I've declared the PR as ready to merge from its previous draft status (which is something I have not done in previous PRs). All requested changes have been already addressed.

DiSanchz committed:
rating_scale_appendices

DiSanchz commented:
I just pushed the last addition discussed in our last meeting on 22/03. That is, a new appendix to both the Marketplace and Non-marketplace tutorials mentioning the existence of the rating scale simulator and briefly explaining its most relevant features. With this, the PR is ready to merge.

Let me know any piece of feedback, suggestions, or corrections you may have to improve these additions before merging! 

@narendramukherjee 
@shrabasteebanerjee 

PD: for further clarity, these additions are Appendix 1 of the non-marketplace simulation tutorial and Appendix 3 of the marketplace simulation tutorial.

DiSanchz committed:
rating_scale_appendices
